smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/action/MetaDataAction.java

```diff
@@ -98,97 +85,105 @@ protected void execute() throws Exception {
-              new Path(srcFile),
-              fs.getFileStatus(new Path(srcFile)).getOwner(),
-              fileInfo.getGroup());
-        }
-        if (fileInfo.getBlockReplication() != -1) {
-          fs.setReplication(new Path(srcFile), fileInfo.getBlockReplication());
-        }
-        if (fileInfo.getPermission() != -1) {
-          fs.setPermission(new Path(srcFile), new FsPermission(fileInfo.getPermission()));
-        }
-        if (fileInfo.getAccessTime() != -1) {
-          fs.setTimes(
-              new Path(srcFile),
-              fs.getFileStatus(new Path(srcFile)).getModificationTime(),
-              fileInfo.getAccessTime());
-        }
-        if (fileInfo.getModificationTime() != -1) {
-          fs.setTimes(
-              new Path(srcFile),
-              fileInfo.getModificationTime(),
-              fs.getFileStatus(new Path(srcFile)).getAccessTime());
-        }
-        return true;
-      } else {
-        // change file metadata in local cluster
-        if (fileInfo.getOwner() != null) {
-          dfsClient.setOwner(
-              srcFile, fileInfo.getOwner(), dfsClient.getFileInfo(srcFile).getGroup());
-        }
-        if (fileInfo.getGroup() != null) {
-          dfsClient.setOwner(
-              srcFile, dfsClient.getFileInfo(srcFile).getOwner(), fileInfo.getGroup());
-        }
-        if (fileInfo.getBlockReplication() != -1) {
-          dfsClient.setReplication(srcFile, fileInfo.getBlockReplication());
-        }
-        if (fileInfo.getPermission() != -1) {
-          dfsClient.setPermission(srcFile, new FsPermission(fileInfo.getPermission()));
-        }
-        if (fileInfo.getAccessTime() != -1) {
-          dfsClient.setTimes(
-              srcFile,
-              dfsClient.getFileInfo(srcFile).getModificationTime(),
-              fileInfo.getAccessTime());
-        }
-        if (fileInfo.getModificationTime() != -1) {
-          dfsClient.setTimes(
-              srcFile,
-              fileInfo.getModificationTime(),
-              dfsClient.getFileInfo(srcFile).getAccessTime());
-        }
-        return true;
+        changeRemoteClusterFileMetadata(srcFile, fileInfoDiff, configuration);
+        return;
       }
-    } catch (Exception e) {
-      LOG.debug("Metadata cannot be applied", e);
+      changeLocalClusterFileMetadata(srcFile, fileInfoDiff, dfsClient);
+    } catch (
+        Exception exception) {
+      LOG.error("Metadata cannot be applied", exception);
+      throw exception;
+    }
+  }
+
+  private static void changeRemoteClusterFileMetadata(String srcFile,
+      FileInfoDiff fileInfoDiff, Configuration configuration) throws IOException {
+    // change file metadata in remote cluster
+    FileSystem fs = FileSystem.get(URI.create(srcFile), configuration);
+    Path srcPath = new Path(srcFile);
+    FileStatus srcFileStatus = fs.getFileStatus(srcPath);
+
+    String owner = srcFileStatus.getOwner();
+    String group = srcFileStatus.getGroup();
```

Comment:
Consider refactoring to reuse the code block that is duplicated in this method: changeLocalClusterFileMetadata.

Reply:
Replaced local `dfsClient` with on the fly constructed local fs and combined methods.